### PR TITLE
Replace the widgets screen experiment with a theme support

### LINKED ISSUE #24685: Introduce the new Widgets editor to the world
### Introduction

Currently as part of Gutenberg's phase 2 process there is an [effort](https://github.com/WordPress/gutenberg/projects/27) put into upgrading the current widgets screen to a block based editor. With the latest progress in #24290 the project is in an architecturally stable point.

It is a good moment to remove the experimental cloak and have it available for a larger user base via the Gutenberg plugin. Issues such as the big list in #24530 need more participation and any help is super valuable. The new UI and UX explorations, even while keeping the same general flow to the current screen will make great use from having the editor exposed to real world projects.

A future step is that after the Widgets  editor is out of being just an experiment, it becomes beta software and the project's focus will be on fixing it, including the default workflow it will offer, and therefore move on with replacing the current screens. 

Aside from using blocks as the base of the UX of managing widgets this project allows users to bring blocks into their widget areas without any required opt in from themes. Testing this important feature via the Gutenberg plugin sooner rather than later is essential for it to work properly.

### Actions

- [ ] Remove the experimental flags that hide the Widgets editor
- [ ] Have the Widgets editor show up under Appearance > Widgets( beta ) in WordPress Admin (for people having the Gutenberg plugin installed)
- [ ] Add a notification in the current widgets screen with an invitation to test the new editor
- [ ] Release the Widgets editor as an optional Appearance link in the next Gutenberg release (8.9)
- [ ] Add the option to restore the current widgets screen and to switch between editors in the Classic Editor plugin
- [ ] Release the Widgets editor as a default Appearance > Widgets link in the next Gutenberg release (9.0)

### PR DESCRIPTION
**Pull request by youknowriad**
closes #24685

The idea at the moment is to try to ship the new block-based widgets screen in 5.6. If we want to make this a reality, we need more testing and a more stable implementation.

This PR does a few things to help with that:

 - Replace the experiments flag with a theme support flag to allow themes to opt-in into the block-based widgets screen.
 - Hides and replaces the existing widgets pages and panel with the block-based ones when theme support is enabled.

### CODE REVIEW

File: gutenberg.php
Comment (youknowriad) [inline gutenberg.php:188]: The support is enabled by default in the Gutenberg plugin to bring visibility to the feature and get more testing, but in Core it could be made opt-in instead of opt-out.

**Review by draganescu — APPROVED** [review]
I like this idea! I tested it locally and it works great.
I am approving but a second opinion might be worth it considering we're hijacking the widgets link in the Gutenberg plugin to point to an experimental screen.

Going back to the old screen would mean I have to disable the Gutenberg plugin?

**Comment by youknowriad** [comment]
> Going back to the old screen would mean I have to disable the Gutenberg plugin?

Alternatively, you could do

```
remove_theme_support( 'block-based-widgets' );
```

File: lib/customizer.php
Comment (mcsf) [inline lib/customizer.php:78]: > if block based widgets are enabled

Seems we need to add this condition inside the function?

Comment (youknowriad) [reply lib/customizer.php:78]: 😬

**Comment by mapk** [comment]
Requiring the theme to opt-in for this feels like a safe step in the right direction, especially if we're replacing the Widgets screen entirely.

**Without theme support**

Screen Shot 2020-07-21 at 9 50 05 AM

**With theme support**

Screen Shot 2020-07-21 at 9 58 03 AM

**Comment by noisysocks** [comment]
One problem with this is that if a user switches away from a theme that supports block widgets, then goes to the (old) widgets screen and clicks _Save_, they will lose any blocks that were stored. See https://github.com/WordPress/gutenberg/issues/24267.

A question I have about defaulting to "opt-out" is: What about old themes that rarely receive updates? I imagined that a requirement of the _blocks in widgets_ project was that it would let users add blocks to their site no matter what theme they're on.

**Comment by youknowriad** [comment]
> I imagined that a requirement of the blocks in widgets project was that it would let users add blocks to their site no matter what theme they're on.

It's still a question to be addressed but I have doubts that we'll be able to make it work regardless of the theme and widgets used thus the opt-in choice.

> One problem with this is that if a user switches away from a theme that supports block widgets, then goes to the (old) widgets screen and clicks Save, they will lose any blocks that were stored. See #24267.

This is a good question and one that we'd need to solve regardless of how the screen is enabled. I'd like for this feature to get more exposure and for us to give more priority to solving these issues.

**Comment by draganescu** [comment]
>  let users add blocks to their site no matter what theme they're on.
This is a nice idea, but, just like with navigation, there is no way to guarantee that it will work. It's does advance the desirable outcome of using blocks with no updates but if the result is some ugly hotchpotch on the front end it will do more harm than good to this outcome. My point is that as long as people can add _any_ block the theme should be tweaked to support that.

In explorations with @adamziel we wondered if the path to the widget editor is reversible. Maybe as @mtias suggested we could add in the classic editor plugin, or in some other plugin, the possibility to revert to the old screens (both for navigation and for widgets). However, are they mutually exclusive? In my opinion they should be, as in if you switch back there should be no guarantee that you can easily continue where you left off. However @adamziel suggested in this thread, that by following the current data structure there won't be any issue to just move between the two screens at any moment, which is a nice to have.

I like this PR because it allows us to put these new editors in the hands of community developers before they're ready to be shipped with the plugin. They could test this and see where things can be improved. It also does not change anything about the future when we merge the new editors: when that happens the opt in is NOT to get the new editor, the opt in is to enable the new editor to _allow_ blocks next to menus or widgets. This PR is only interesting at this stage. After the merge everyone will get the new editors but only the opting in themes will allow blocks to be used.

**Comment by TimothyBJacobs** [comment]
It'd be nice to also use `register_theme_feature` when available.

File: gutenberg.php
Comment (mtias) [reply gutenberg.php:191]: Can we rename it to something like "widgets-block-editor"?

**Comment by youknowriad** [comment]
Merging this based on the last discussions on #core-editor chat.

**Comment by mapk** [comment]
💥

**Pull request merged**
